pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/AggregationUtils.java

```diff
@@ -54,54 +54,92 @@ public static Key extractEmptyKey() {
     return new Key(new Object[0]);
   }
 
-  private static Object mergeSum(Object left, Object right) {
-    return ((Number) left).doubleValue() + ((Number) right).doubleValue();
+  // TODO: Use the correct type for SUM/MIN/MAX instead of always using double
+
+  @Nullable
+  private static Object mergeSum(@Nullable Object agg, @Nullable Object value) {
+    if (agg == null) {
+      return value;
+    }
+    if (value == null) {
+      return agg;
+    }
+    return ((Number) agg).doubleValue() + ((Number) value).doubleValue();
   }
 
-  private static Object mergeMin(Object left, Object right) {
-    return Math.min(((Number) left).doubleValue(), ((Number) right).doubleValue());
+  @Nullable
+  private static Object mergeMin(@Nullable Object agg, @Nullable Object value) {
+    if (agg == null) {
+      return value;
+    }
+    if (value == null) {
+      return agg;
+    }
+    return Math.min(((Number) agg).doubleValue(), ((Number) value).doubleValue());
   }
 
-  private static Object mergeMax(Object left, Object right) {
-    return Math.max(((Number) left).doubleValue(), ((Number) right).doubleValue());
+  @Nullable
+  private static Object mergeMax(@Nullable Object agg, @Nullable Object value) {
+    if (agg == null) {
```

Comment:
Just curious, today the leaf node leverages the nullValueBitmap, but this is not available in the intermediate stages.

Example:
```
    if (_nullHandlingEnabled) {
      // TODO: avoid the null bitmap check when it is null or empty for better performance.
      RoaringBitmap nullBitmap = blockValSet.getNullBitmap();
      if (nullBitmap == null) {
        nullBitmap = new RoaringBitmap();
      }
      aggregateNullHandlingEnabled(length, aggregationResultHolder, blockValSet, nullBitmap);
      return;
    }
```

Now we usually split the LogicalAggregate into an intermediate stage (which may or may not run on leaf) and a final stage (which probably won't run on leaf as it needs to do the final aggregations). Does this mean that the nullValueBitmap won't be leveraged if this intermediate stage doesn't run in the leaf? will that affect query results?

Reply:
try to piece out what this comment is about, here is my understanding:

- intermediate stage is run on `List<Object[]>` so it can naturally represent null with the object array
- what you mentioned with nullValueBitmap is being used in 2 places
    - in leaf operators that will read it out when `enableNullHandling` flag is turned on
    - in data table ser/de --> this also means when intermediate stage received the payload, it will deserialized it into `List<Object[]>` with the null put right into places. 

please let me know if this answers your question

Reply:
Thanks for getting back on this! So to make sure I understand correctly, when the data is sent from leaf to intermediate, the null blocks will be correctly set to null in the `List<Object[]>`? Thus making a simple null check like this good enough?

Reply:
Correct. If null handling is enabled, `null` should be properly set into the `Object[]`, and the null bitmap is already dropped

Reply:
> when the data is sent from leaf to intermediate, the null values will be correctly set to null in the `List<Object[]>`? Thus making a simple null check like this good enough?

this statement is not entirely true. 
- when sent, the data is still Ser/De into bitmap (see `DataBlock`)
- when used, it is pulled from `TransferableBlock` which has 2 member variable format: (1) `DataBlock _dataBlock` format and (2) `List<Object[]> _container`

it is when the lazy eval of the _container that puts the null in the right place

Reply:
Got it, thanks for explaining in detail! Checking out that part of the code